### PR TITLE
Differences in struct layout between memory and storage

### PR DESCRIPTION
**Pull request by ritzdorf**
More explicit description with an example.

### CODE REVIEW

File: docs/internals/layout_in_memory.rst [hunk 1]
```diff
@@ -20,6 +20,17 @@ memory is never freed (this might change in the future).
 
 Elements in memory arrays in Solidity always occupy multiples of 32 bytes (this
 is even true for ``byte[]``, but not for ``bytes`` and ``string``).
+This implies that structs can take up more space in memory than they do in
```
Comment (mijovic) [inline docs/internals/layout_in_memory.rst]:
```suggestion
This implies that structs can take up more space in memory than they do in
```

**Comment by mijovic** [comment]
Also here documentation is mentioning this is true for memory arrays, so I think you would need to update your example a little bit.

**Comment by ritzdorf** [comment]
> Also here documentation is mentioning this is true for memory arrays, so I think you would need to update your example a little bit.

I wanted to make an example for structs, as this is a question I have often heard about. Happy to add one for arrays too. Is that what you mean @mijovic ?

**Comment by mijovic** [comment]
> > Also here documentation is mentioning this is true for memory arrays, so I think you would need to update your example a little bit.
> 
> I wanted to make an example for structs, as this is a question I have often heard about. Happy to add one for arrays too. Is that what you mean @mijovic ?

I am not sure if this is best place for example about structs. But, in general this example can be a bit misleading. Storage is divided to slots of 256 bytes. In case of your struct, it will occupy one slot.
The only way to use reminder of that slot is via inline assembly.

But definitely, documentation lacks info on layouts of different types.

I suggest moving this to different location in documentation and using a slight different wording.

**Comment by ritzdorf** [comment]
>  Storage is divided to slots of 256 bytes. 

I guess you meant 256 bit = 32 bytes.

>  In case of your struct, it will occupy one slot.

The struct occupies 3 slots.

> I suggest moving this to different location in documentation and using a slight different wording.

I moved it to a different subsection.

Thanks for your feedback @mijovic .

File: docs/internals/layout_in_memory.rst [hunk 2]
```diff
@@ -36,4 +36,41 @@ elements.
...
+    struct S{
+        uint a;
+        uint b;
+        uint8 c;
+        uint8 d;
+    }
+
+
+
+.. index: calldata layout
```
Comment (chriseth) [inline docs/internals/layout_in_memory.rst]: It seems this line belong to the top of `layout_in_calldata` - can you move it there, please?

File: docs/internals/layout_in_memory.rst [hunk 3]
```diff
@@ -36,4 +36,41 @@ elements.
...
+Example for Difference in Struct Layout
+---------------------------------------
+
+The following struct occupies 96 bytes (3 slots of 32 bytes) in storage,
+but 128 bytes (4 items with 32 bytes each) in memory.
+
+
+::
+
+    struct S{
```
Comment (chriseth) [inline docs/internals/layout_in_memory.rst]:
```suggestion
    struct S {
```

**Comment by mijovic** [comment]
@ritzdorf Sorry for completely off review. I wasn't fully concentrated.
Thanks for moving it into different section. After fixing few comments on this PR from Chris, I guess it would be good to merge.

File: docs/internals/layout_in_memory.rst (hunk 2 above)
Comment (ritzdorf) [reply docs/internals/layout_in_memory.rst]: Thanks @chriseth ! Fixed in 8a2731bfdb32aa81a84d6da3ef211958073bae4a

**Comment by mijovic** [comment]
Can you squash into one commit?

**Comment by ritzdorf** [comment]
> Can you squash into one commit?

Done. @mijovic

**Review by mijovic — APPROVED** [review]

**Review by cameel — APPROVED** [review]

**Pull request merged**